### PR TITLE
Add crowdfund contract and initial tests

### PR DESCRIPTION
**Pull request by iangregsondev**

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -1,46 +1,34 @@
...
-                "-t",
-                "100000"
-            ],
-            "preLaunchTask": "tsc: build - tsconfig.json",
-            "outFiles": []
-        }
-    ]
-}
\ No newline at end of file
+  // Use IntelliSense to learn about possible attributes.
```
Comment (xhliu) [inline .vscode/launch.json]: Please do not change and commit this file.

File: src/contracts/crowdfund.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    super(recepient, contributor, deadline, target);
+    this.recepient = recepient;
+    this.contributor = contributor;
+    this.deadline = deadline;
+    this.target = target;
+  }
+
+  @method()
+  public collect(txPreimage: SigHashPreimage, raisedAmount: bigint) {
+    assert(this.checkPreimage(txPreimage));
```
Comment (xhliu) [inline src/contracts/crowdfund.ts]: You can use `this.ctx.hashOutputs` in the latest version without `SigHashPreimage`, like in https://scrypt.io/scrypt-ts/tutorials/stateful-contract.

File: src/contracts/crowdfund.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      new bsv.Transaction.Output({
+        script: this.lockingScript,
+        satoshis: initBalance,
+      })
+    );
+    this.lockTo = { tx, outputIndex: 0 };
+    return tx;
+  }
+
+  //   getCallTxForCollect(
```
Comment (xhliu) [inline src/contracts/crowdfund.ts]: Remove dead code everywhere.

**Review by xhliu — CHANGES_REQUESTED** [review]
👍 awesome.

Please fix the failing CI.

File: src/contracts/crowdfund.ts (hunk 2 above)
Comment (iangregsondev) [reply src/contracts/crowdfund.ts]: Don't understand that I am afraid. Seems like I need to use a stateful contract and this is not a stateful contract.

File: src/contracts/crowdfund.ts (hunk 3 above)
Comment (iangregsondev) [reply src/contracts/crowdfund.ts]: Done

**Comment by xhliu** [comment]
```ts
    // pledge fund
    @method
    public pledge(int raisedAmount) {
        // reach target
        asssert(raisedAmount >= this.target);

        // fund goes to the recipient
        bytes lockingScript = Utils.buildPublicKeyHashScript(this.recepient);
        
        bytes output = Utils.buildOutput(lockingScript, raisedAmount);
        assert(hash256(output) == this.ctx.hashOutputs);

    }

    // contributor can take the fund back after deadline
   @method
    public refund(Sig sig) {
        // fundraising expired
        assert(this.ctx.locktime >= this.deadline);

        assert(this.checkSig(sig, this.contributor));
    }
```

**Comment by iangregsondev** [comment]
> ```ts
>     // pledge fund
>     @method
>     public pledge(int raisedAmount) {
>         // reach target
>         asssert(raisedAmount >= this.target);
> 
>         // fund goes to the recipient
>         bytes lockingScript = Utils.buildPublicKeyHashScript(this.recepient);
>         
>         bytes output = Utils.buildOutput(lockingScript, raisedAmount);
>         assert(hash256(output) == this.ctx.hashOutputs);
> 
>     }
> 
>     // contributor can take the fund back after deadline
>    @method
>     public refund(Sig sig) {
>         // fundraising expired
>         assert(this.ctx.locktime >= this.deadline);
> 
>         assert(this.checkSig(sig, this.contributor));
>     }
> ```

Thanks, I have updated it.

Seems logical now :-)

But if I don't pass in sighash types, then the hashed output is different, if I pass in the sighash types then it fails silently.

**Comment by iangregsondev** [comment]
You will notice that I am passing the following, but this does fail silently, in fact on a debug session the "pledge" method is not even entered,

```
      .setInputScript(
        {
          inputIndex,
          sigtype:
            bsv.crypto.Signature.SIGHASH_ANYONECANPAY |
            bsv.crypto.Signature.SIGHASH_SINGLE |
            bsv.crypto.Signature.SIGHASH_FORKID,
        },
        (tx) => {
          this.unlockFrom = { tx, inputIndex };
          return this.getUnlockingScript((self) => {
            self.pledge(1000n);
          });
        }
      );
````

**Comment by iangregsondev** [comment]
I did also try SIGHASH_ALL in place of SINGLE - but the same result

**Comment by iangregsondev** [comment]
If my understanding is correct, the SIGHASH_ANYONECANPAY must be passed, because otherwise the inputs can't be changed.

**Comment by iangregsondev** [comment]
sigtype:
            bsv.crypto.Signature.SIGHASH_NONE |
            bsv.crypto.Signature.SIGHASH_ANYONECANPAY |
            bsv.crypto.Signature.SIGHASH_FORKID,

**Comment by xhliu** [comment]
U do not need ANYONECANPAY here, since there is no `checkSig` in `pledge`.

The default should work.
`ALL | FORKID`

**Comment by iangregsondev** [comment]
Ah, ok, thanks.

I have removed it and it does work but the tests fail because the hash is different

I.e.

```
Test SmartContract `Crowdfund`
compiling contract /home/ian/Development/scrypt/forks/scryptTS-examples/scrypts/src/contracts/crowdfund.scrypt ...
here is the lockingscript  76a914528223920ada5975f41c217061801b558b508ce288ac
here is the output  e8030000000000001976a914528223920ada5975f41c217061801b558b508ce288ac
here is the hashoutputs  6cbad773d0124aa4481b141bab4e1ac31aa970a1b8e9e1f40240920dd699d2b0
here is the hash256(output)  93f43d9b2c54ec1455190b7d40136ab5b0cb07e38800aa778549046802d9216f
    1) should pass the public method unit test successfully.

```

File: src/contracts/crowdfund.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      .setOutput(0, (tx: bsv.Transaction) => {
+        nextInst.lockTo = { tx, outputIndex: 0 };
+        return new bsv.Transaction.Output({
+          script: nextInst.lockingScript,
+          satoshis: 1000,
+        });
+      })
+      .setInputScript(
+        {
+          inputIndex,
```
Comment (iangregsondev) [inline src/contracts/crowdfund.ts:121]: Removed sigTypes here

Comment (iangregsondev) [reply src/contracts/crowdfund.ts:121]: Also tried adding

```
          sigtype:
            bsv.crypto.Signature.SIGHASH_ALL |
            bsv.crypto.Signature.SIGHASH_FORKID,

```

Comment (iangregsondev) [reply src/contracts/crowdfund.ts:121]: These do enter the method now, but the hash again fails as they are not the  same - (last assert)

**Pull request closed without merging**